### ISSUE TITLE
`<For />`:  when removing elements, getting "panicked at 'called `Option::unwrap()` on a `None` value'" at each.rs:853, in some conditions

### ISSUE THREAD

**Issue by nguiard** [leptos-rs/leptos#1274]
**Describe the bug**

When clearing multiple elements of a Vec used in a `<For />`, for some combinations of removed elements, I get 
```
start-axum.js:569 panicked at 'called `Option::unwrap()` on a `None` value', [...]/leptos_dom/src/components/each.rs:853:57
```

I have a hard time producing a simple reproduction for now, working on it... But as it seems to point to a specific line in leptos, I thought you might have an idea.

In my app (still trying to make a nice todo app), the problem happens only when removing some specific subsets of elements. 
- with 3 elements in the list, I'm unable to produce this message
- with 4 elements in the list however (indexes [0, 1, 2, 3]), the error happens only when removing elements [0, 2] (at the same time, using `Vec::retain`)
- with 5 elements in the list, it seems to happens with a lot of combinations, apparently *all* of these involving removing the first element (index 0) *and not* removing the second (index 1). But also not in all of the situations involving these conditions produce the bug... (removing indices [0, 2, 3, 4] works fine for example, but not removing [0, 2, 4])

As I said I have a hard time producing a minimal example right now, but I'll work on it when I can...
Also, I *can* work around this issue by looping over the elements in the vec, and calling `update` sequentially for each one I want to remove, removing only this one (but yeah, this ain't very pretty nor declarative, and probably not performant at all)...

In any case, thanks for the great work on leptos, loving it :)

Edit: sorry, I forgot to mention, I'm on the latest commit at the time of writing (66f54e7)

**Comment by godstail** [comment]
I also had this panic, but in my case it was when one of the item's key wasn't stable and got changed with each rerender.
Slightly different line number in my case, 849 (prob. different version):
```let each_item = children[move_.from].take().unwrap();```

**Comment by nguiard** [comment]
Hmm interesting... In my case, as far as I can tell, the items' keys are actually stable...

**Comment by gbj** [comment]
@jquesada2016 This is caused by a situation in which the diff causes multiple moves. 

Going from `[0, 1, 2, 3]` to `[1, 3]` yields the diff
```
Diff {
    removed: [
        DiffOpRemove {
            at: 0,
        },
        DiffOpRemove {
            at: 2,
        },
    ],
    moved: [
        DiffOpMove {
            from: 1,
            len: 2,
            to: 0,
            move_in_dom: false,
        },
    ],
    items_to_move: 2,
    added: [],
    clear: false,
}
```

After `unpack_moves`, `move_cmds` is 
```
[
    DiffOpMove {
        from: 1,
        len: 1,
        to: 0,
        move_in_dom: false,
    },
    DiffOpMove {
        from: 2,
        len: 1,
        to: 1,
        move_in_dom: false,
    },
]
```

and `children` is 
```
[
    None,
    Some(
        <EachItem>
            <span>1</span>
        </EachItem>,
    ),
    None,
    Some(
        <EachItem>
            <span>3</span>
        </EachItem>,
    ),
]
```

So when this runs
```rust
    let mut moved_children = move_cmds
        .iter()
        .rev()
        .map(|move_| {
            let each_item = children[move_.from].take().unwrap();

            if move_.move_in_dom {
                each_item.prepare_for_move();
            }

            Some(each_item)
        })
        .collect::<Vec<_>>();
```
The first one is fine (takes from index 1). The second one tries to take from index 2 — but the item is really at index 3, and index 2 is empty, so panics.

Should that second unpacked move actually be from `3` to `1`?

**Comment by jquesada2016** [comment]
@gbj, thanks for the detailed breakdown. I'll read this more thoroughly tomorrow as soon as I get a chance to working on fixing this, but from skimming over this, it seems like this is due to the `Diff` returning `items_to_move: 2` rather than `1`, so it might just be doing an extra iteration in the `unpack` function.

**Comment by gbj** [comment]
This is fixed by #1296.
